README.md
# Unsafe API

This is a personal project for creating local API backends for automated tasks.
It is not intended to be used in production environments.

Main entity "Item" can be used for testing CRUD operations on MongoDB.

## Quick start
Copy `.env-sample` to `.env`, edit configuration.
Run following commands:
```
   npm install
   nodemon server
```

### Check status
http://localhost:4200/api/status

### Add item
POST http://localhost:4200/api/item
```
{
  "name": "hello world",
}
```

## Run Server as a Service

```
npm install pm2 -g
pm2 start server
```


history.md
npm init

git init

npm i dotenv express mongoose morgan cors helmet hammerjs joi body-parser compression cookie-parser express-async-handler

npm i nodemon --save-dev


server/config/config.js
const Joi = require('joi');

// require and configure dotenv, will load vars in .env in PROCESS.ENV
require('dotenv').config();

// define validation for all the env vars
const envVarsSchema = Joi.object({
  NODE_ENV: Joi.string()
    .allow('development', 'production', 'test', 'provision')
    .default('development'),
  SERVER_PORT: Joi.number()
    .default(4040),
  MONGOOSE_DEBUG: Joi.boolean()
    .when('NODE_ENV', {
      is: Joi.string().equal('development'),
      then: Joi.boolean().default(true),
      otherwise: Joi.boolean().default(false)
    }),
  MONGO_HOST: Joi.string().required()
    .description('Mongo DB host url'),
  MONGO_PORT: Joi.number()
    .default(27017)
}).unknown()
  .required();

const { error, value: envVars } = envVarsSchema.validate(process.env);
if (error) {
  throw new Error(`Config validation error: ${error.message}`);
}

const config = {
  env: envVars.NODE_ENV,
  port: envVars.SERVER_PORT,
  mongooseDebug: envVars.MONGOOSE_DEBUG,
  mongo: {
    host: envVars.MONGO_HOST,
    port: envVars.MONGO_PORT
  }
};

module.exports = config;


server/controllers/item.controller.js
const Joi = require('joi');
const Items = require('../models/item.model');

exports.index = function (req, res) {

  console.log( req.body )
  console.log( req.query )

  var search = { active: true };
  projection = 'name created updated'
  limit = 10
  orderby = 'updated'
  order = '1'

  Items.find(
    search,
    projection,
    function( err, items ) {
      if (err) { 
        res.json({
          status: "error",
          message: err,
        });
      } else {

        // count of results
        Items.countDocuments( search, function(err, numberOfItems) {
          res.json({
            status: "success",
            message: "Items retrieved successfully.", 
            data: items,
            count: numberOfItems
          });
        });

      }
    })
    .sort({[orderby]: order})
    .limit(limit);
};

async function insert(req, res){

  if( !('name' in req.body ) ){
    res.status(400).json({
      status: "error",
      message: "name is required"
    })
  }

  item = req.body

  newItem = new Items( item )
  newItem.save(function (err, inserted ) {
    if (err){
      return res.status(500).json({
        status: "error",
        message: err
      })
    } else {
      return res.status(201).json({
        status: "ok",
        message: "New item is created",
        data: inserted.id
      })
    }
  })
};

exports.view = function (req, res) {
  Items.findById(req.params.item_id, function (err, item) {
    if (err){
      res.send(err);
    } else {
      res.json({
        message: 'Item details are loaded.',
        data: item
      });
    }
  })
};

async function update(req, res) {
  Items.findById( req.params.item_id, async function ( err, item ) {
    if (err) {
      res.send(err);
    } else {

      item.name = req.body.name

      item.updated = Date.now(); 

      item.save(function (err) {
        if (err) {
          res.json(err);
        } else {
          res.json({
            message: 'Item updated.',
            data: item
          });
        }
      });
    }
  });
};

exports.delete = function (req, res) {
  Items.findById(req.params.item_id, function (err, item) {
    if (err) {
      res.send(err);
    } else {

      item.active = false
      item.updated = Date.now(); 

      item.save(function (err) {
        if (err) {
          res.json(err);
        } else {
          res.json({
            message: 'Item deleted.',
            data: item
          });
        }
      });
    }
  });
};

async function patch(req, res) {

  Items.findById(req.params.item_id, async function (err, item) {
    if (err) {
      res.send(err);
    } else {

      if( req.body.hasOwnProperty('name') ){
        item.name = req.body.name
      }

      item.updated = Date.now(); 

      item.save(function (err) {
        if (err) {
          res.json(err);
        } else {
          res.json({
            message: 'Item updated.',
            data: item
          });
        }
      });
    }
  });
};


// exports
exports.insert = insert
exports.update = update
exports.patch = patch


server/config/express.js

const express = require( 'express' );
const path = require( 'path' );
const httpError = require( 'http-errors' );

const helmet = require( 'helmet' );
const logger = require( 'morgan' );
const compression = require( 'compression' );
const cookieParser = require( 'cookie-parser' );

const config = require( './config' );
const routes = require( '../routes/index.route' );

const app = express();

if( config.env === 'development' ){
  app.use( logger('dev') );
  console.log( 'Morgan is enabled.' );
}

var distDir = '../../dist';
app.use( express.static( path.join( __dirname, distDir ) ) );
app.use( /^((?!(api)).)*/, (req, res) => {
  res.sendFile( path.join( __dirname, distDir, '/index.html' ) );
});

app.use( express.json({limit: '50mb'}) );
app.use( express.urlencoded( { extended: true, limit: '50mb'} ) );

app.use( compression() );
app.use( cookieParser() );
app.use( helmet() );

app.use( '/api/', routes );

app.use( (req, res, next) => {
  const err = new httpError(404);
  return next(err);
} );

// Error handling
app.use( (err, req, res, next) => {

  if (err.isJoi) {
    err.message = err.details.map( e => e.message ).join("; ");
    err.status = 400;
  }

  res.status( err.status || 500 ).json({ status: "error", message: err.message });
  next(err);

});

module.exports = app;